### PR TITLE
Milestone 5: Private-events

### PR DESCRIPTION
**Pull request by enaburekhan**
**Private events built with Ruby on rails.**
**In this project:**

- Create associations between User, Event and Attendance
- Create logic in view to separate upcoming events from past events
- Create sessions to cache user_id on the browser
- Use bootstrap for styling

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class EventsController < ApplicationController
+  include EventsHelper
+  before_action :authorization
+  helper_method :events_user_attending, :test
+
+  def index
+    @event = Event.all
+    @past_events = Event.past.paginate(page: params[:page], per_page: 2).order(created_at: :desc)
+    @upcoming_events = Event.upcoming.paginate(page: params[:page], per_page: 2).order(created_at: :desc)
```
Comment (voscarmv) [inline app/controllers/events_controller.rb:9]: Please move the `paginate` and `order` preprocessing to the scope methods instead of writing them here.

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far, team. 👍 

## However, there are some changes required 👀 🧰 🔨 🔧 

- Please add validation to the event name so that it's not possible to create events with an empty name.
- Please add validation to the event date so that it's required for the event to have a date. (the date can not be left empty)
- Currently there is no way to attend events in the app. Please add a button or some sort of input where a user can attend an event.
- Please make sure the Event belongs to User and has a foreign key with `has_many :events, foreign_key: :creator_id`
- Please read my code comments for further necessary changes.

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Eric Enaburekhan & Ceejay Okoli** 
Good job with the improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to correct before you can go ahead:

- [ ] User's should be able to attend an event
- [ ] Remove commented code in User class
- [ ] Remove commented code in Attendance
 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: app/views/layouts/_nav.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            <%= link_to 'Create Event', new_event_path(@event), class: "nav-link"%>
+        </li>
+        <li class="nav-item">
+            <%= link_to 'Logout', logout_path(@user), class: "nav-link", method: :delete %>
+        </li>
+      <% else %>
+        <li class="nav-item">
+            <%= link_to 'Sign up', new_user_path(@user), class: "nav-link" %>
+        </li>
+      <%end%>
```
Comment (teekaytech) [inline app/views/layouts/_nav.html.erb]: This logic can and should be separated into a helper file.

File: app/views/users/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          <h2 class="title is-2 is-spaced has-text-centered mb-6">Guest List</h2>
+    
+          <div class="columns is-centered">
+            <table class="table is-fullwidth is-hoverable">
+              <tbody>
+                <th class="has-text-centered">Name</th>
+                <th class="has-text-centered">Status</th>
+                <th class="has-text-centered">Action</th>
+                <% @user.reject { |user| user == Event.find(params[:event_id]).creator }.each do |user| %>
+                  <% attendance = Attendance.find_by(event_id: params[:event_id], user_id: user.id) %>
```
Comment (teekaytech) [inline app/views/users/index.html.erb]: `Attendance.find_by` query should have been done in the users_controller#index action.

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,51 @@
+<div class="section">
+  <div class="container">
+	  <div>
+        <h2 class="title is-2 is-spaced has-text-centered"><%= @user.name %>'s Profile</h2>
+      <div>
+	      <div class="row">
+          <div class="col-4">
+            <h3 class="subtitle is-3 has-text-centered">Hosted Events: <%= @user.events.count %></h3>
+            <ul>
+              <% if @user.events.empty? %>
```
Comment (teekaytech) [inline app/views/users/show.html.erb]: @user.events query should be made in the users_controller#show action.

File: app/views/users/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                  <li class="has-text-centered"><%=link_to event.description, event_path(event) %></li>
+                <% end %>
+              <% end %>
+            </ul>
+          </div>
+
+          <div class="col-4">
+            <h3 class="subtitle is-3">Upcoming Events: <%= @user.attended_events.where('date >= ?', Date.today).count %></h3>
+            <ul>
+              <% if @user.attended_events.where('date >= ?', Date.today).count.zero? %>
```
Comment (teekaytech) [inline app/views/users/show.html.erb]: @user.attended_events should also be called in the controller.

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__Your commitment and genuinely in this project is commendable. 
You just have some changes to make before this project can be approved.__

**[ REQUIRED CHANGES ]**

- [ ] There should be NO application logic in the views nor helpers (no `Post.new, Comment.create, current_user.friends.find_by(...)...`). Everything connected with data processing should be encapsulated in models and used in controllers.
- Check my inline comments for details.

- [ ] The event's show page displays the list of attendants 
- Right now, the event show page only displays a number of attendants. Since the requirement wants you to display the list of attendants, kindly refactor your code to welcome this development.

- [ ] You still have some commented code in user.rb and attendance.rb


* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Comment by enaburekhan** [comment]
**Code review changes effected:**

- The event_show view shows the list of attendees as shown in the image below:

attendee

- refactor codes with logic in view to helper.

**Review by Dandush03 — APPROVED** [review]
# Status :rocket: Approved :rocket: 
### Great Work Guys!
# Happy Coding :+1:

**Pull request merged**